src/macaron/repo_finder/repo_finder.py

```diff
@@ -148,3 +132,129 @@ def to_repo_path(purl: PackageURL, available_domains: list[str]) -> str | None:
+        logger.debug("PURL is missing version.")
+        return False
+
+    found_repo = input_repo
+    if not input_repo:
+        logger.debug("Searching for repo of PURL: %s", purl)
+        found_repo = find_repo(purl)
+
+    if not found_repo:
+        logger.error("Could not find repo for PURL: %s", purl)
+        return False
+
+    # Disable other loggers for cleaner output.
+    logging.getLogger("macaron.slsa_analyzer.analyzer").disabled = True
+    logging.getLogger("macaron.slsa_analyzer.git_url").disabled = True
+
+    if defaults.getboolean("repofinder", "find_source_should_clone"):
+        logger.debug("Preparing repo: %s", found_repo)
+        repo_dir = os.path.join(global_config.output_path, GIT_REPOS_DIR)
+        git_obj = prepare_repo(
+            repo_dir,
+            found_repo,
+            purl=purl,
+        )
+
+        if not git_obj:
+            # TODO expand this message to cover cases where the obj was not created due to lack of correct tag.
+            logger.error("Could not resolve repository: %s", found_repo)
+            return False
+
+        try:
+            digest = git_obj.get_head().hash
+        except ValueError:
+            logger.debug("Could not retrieve commit hash from repository.")
+            return False
+    else:
+        # Retrieve the tags.
+        tags = get_tags_via_git_remote(found_repo)
+        if not tags:
+            return False
+
+        matches = match_tags(list(tags.keys()), purl.name, purl.version)
+
+        if not matches:
+            return False
+
+        matched_tag = matches[0]
+        digest = tags[matched_tag]
+
+    if not digest:
+        logger.error("Could not find commit for purl / repository: %s / %s", purl, found_repo)
+        return False
+
+    if not input_repo:
+        logger.info("Found repository for PURL: %s", found_repo)
+
+    logger.info("Found commit for PURL: %s", digest)
+
+    if not generate_report(purl_string, digest, found_repo, os.path.join(global_config.output_path, "reports")):
+        return False
+
+    return True
+
+
+def get_tags_via_git_remote(repo: str) -> dict[str, str] | None:
+    """Retrieve all tags from a given repository using ls-remote.
+
+    Parameters
+    ----------
+    repo: str
+        The repository to perform the operation on.
+
+    Returns
+    -------
+    dict[str]
+        A dictionary of tags mapped to their commits, or None if the operation failed..
+    """
+    tags = {}
+    try:
+        tag_data = git.cmd.Git().ls_remote("--tags", repo)
+    except git.exc.GitCommandError as error:
+        logger.debug("Failed to retrieve tags: %s", error)
+        return None
+
+    for tag_line in tag_data.splitlines():
+        tag_line = tag_line.strip()
+        if not tag_line:
+            continue
+        split = tag_line.split("\t")
+        if len(split) != 2:
+            continue
+        possible_tag = split[1]
+        if "{" in possible_tag or "}" in possible_tag:
```

Comment:
I think it would be good to document an example of this scenario. 
I'm curious why `{` or `}` appear in the result 🤔. Do you happen to know the reason behind it?

Reply:
I have not found the reason for it. If we take `avaje-prism` as an example the list of reported tags includes:
```
d432a93941b98e4879a99fa294c5b98501c1bc54        refs/tags/avaje-prisms-parent-1.3
db60daeba004c446fbf8e98cacd7f062f47d3985        refs/tags/avaje-prisms-parent-1.3^{}
```
In this case the above snippet may actually be wrong, as if we look at the GitHub UI for tags we find that the commit of the tag `avaje-prism-parent-1.3` is the one of the tag ending with `^{}`. See https://github.com/avaje/avaje-prisms/tags?after=1.6

Reply:
Further investigation reveals that tags ending with `^{}` are what the tag is pointing to. Also known as "peeled tags". If they exist, they should be chosen instead so I'll update the code to reflect that.

Reply:
Right so from that example `d432a93941b98e4879a99fa294c5b98501c1bc54` is the hash of the annotated tag object. That tag objects references the commit object of `db60daeba004c446fbf8e98cacd7f062f47d3985`